File: src/javascript/algorithms/array/10.maxSumCircularSubarray/maxSumCircularSubarray.js
function maxSumCircularSubarray(nums) {
    if(nums.length === 0) return 0;

    let globalMaxSum = nums[0], globalMinSum = nums[0];
    let currentMaxSum = 0, currentMinSum = 0;
    let totalSum = 0;

    for (const num of nums) {
        currentMaxSum = Math.max(currentMaxSum + num, num);
        globalMaxSum = Math.max(currentMaxSum, globalMaxSum);

        currentMinSum = Math.min(currentMinSum + num, num);
        globalMinSum = Math.min(currentMinSum, globalMinSum);

        totalSum += num;
    }

    // If all numbers are negative, total - globalMinSum would be 0 or less
    return globalMaxSum > 0 ? Math.max(globalMaxSum, (totalSum-globalMinSum)) : globalMaxSum;
}

// Test Cases
// ---------------------------
const testCases = [
    { input: [9, -9, 6, 11, -6, -10, 15, 1], expected: 'Varies' },
    { input: [6, -2, 6], expected: 12 },
    { input: [-6, -2, -6], expected: -2 },
    { input: [], expected: 0 }
];

for (const { input, expected } of testCases) {
    const result = maxSumCircularSubarray(input);
    console.log(`Input: ${JSON.stringify(input)} => Max Circular Subarray Sum: ${result} (Expected: ${expected})`);
}

File: src/python/algorithms/array/10.maxSumCircularSubarray/max_sum_circular_subarray.py
"""
Maximum Sum Circular Subarray

Given a circular integer array nums, return the maximum possible sum
of a non-empty subarray of nums.

Time Complexity: O(n)
Space Complexity: O(1)
"""


def max_sum_circular_subarray(nums):
    """
    Kadane's algorithm tracking both max and min subarrays.
    TC: O(n), SC: O(1)
    """
    if not nums:
        return 0
    
    global_max_sum = nums[0]
    global_min_sum = nums[0]
    current_max_sum = 0
    current_min_sum = 0
    total_sum = 0
    
    for num in nums:
        current_max_sum = max(current_max_sum + num, num)
        global_max_sum = max(current_max_sum, global_max_sum)
        
        current_min_sum = min(current_min_sum + num, num)
        global_min_sum = min(current_min_sum, global_min_sum)
        
        total_sum += num
    
    # If all numbers are negative, total_sum - global_min_sum would be 0
    # So return global_max_sum in this case
    if global_max_sum > 0:
        return max(global_max_sum, total_sum - global_min_sum)
    return global_max_sum


# Test cases
if __name__ == "__main__":
    test_cases = [
        [9, -9, 6, 11, -6, -10, 15, 1],
        [6, -2, 6],
        [-6, -2, -6],
        []
    ]
    
    for nums in test_cases:
        result = max_sum_circular_subarray(nums)
        print(f"Input: {nums} => Max Circular Subarray Sum: {result}")
